parsers/US_PJM.py

```diff
@@ -0,0 +1,135 @@
+#!/usr/bin/env python3
+
+"""Parser for the PJM area of the United States."""
+
+import arrow
+from bs4 import BeautifulSoup
+import demjson
+import re
+import requests
+
+
+url = 'http://www.pjm.com/markets-and-operations.aspx'
+
+mapping = {
+            'Coal': 'coal',
+            'Gas': 'gas',
+            'Hydro': 'hydro',
+            'Multiple Fuels': 'unknown',
+            'Nuclear': 'nuclear',
+            'Oil': 'oil',
+            'Other': 'unknown',
+            'Other Renewables': 'unknown',
+            'Solar': 'solar',
+            'Wind': 'wind'
+            }
+
+
+def extract_data(session = None):
+    """
+    Makes a request to the PJM data url.
+    Finds timestamp of current data and converts into a useful form.
+    Finds generation data inside script tag.
+    Returns a tuple of generation data and datetime.
+    """
+
+    s = session or requests.Session()
+    req = requests.get(url)
+    soup = BeautifulSoup(req.content, 'html.parser')
+
+    try:
+        time_div = soup.find("div", id="asOfDate").text
+    except AttributeError:
+        raise LookupError('No data is available for US-PJM.')
+
+    time_pattern = re.compile(r"""(\d{1,2}     #Hour can be 1/2 digits.
+                                   :           #Separator.
+                                   \d{2})\s    #Minutes must be 2 digits with a space after.
+                                   (a.m.|p.m.) #Either am or pm allowed.""", re.X)
+
+    latest_time = re.search(time_pattern, time_div)
+
+    time_data = latest_time.group(1)
+    am_or_pm = latest_time.group(2)
+    hour = int(time_data.split(":")[0])
+    minute = int(time_data.split(":")[1])
+
+    #Time format used by PJM is slightly unusual and needs to be converted so arrow can use it.
+    if am_or_pm == "p.m." and hour != 12:
+        hour += 12
+
+    arr_dt = arrow.now('America/New_York').replace(hour=hour, minute=minute)
+    dt = arr_dt.floor('minute').datetime
+
+    generation_mix_div = soup.find("div", id="rtschartallfuelspjmGenFuelM_container")
+    generation_mix_script = generation_mix_div.next_sibling
+
+    pattern = r'series: \[(.*)\]'
+    script_data = re.search(pattern, str(generation_mix_script)).group(1)
+    raw_data = demjson.decode(script_data)
```

Comment:
I'm guessing demjson is used because the source doesn't double-quote dict keys and has only single quotes around string values? Maybe add a comment explaining if so.

Reply:
Thanks @systemcatch 

For simplicity's sake...

Realtime exchange flows are here... http://www.pjm.com/markets-and-operations/interregional-map.aspx
            # No. 1 - US-PJM->US-NY
                # to US_NY are combined from PJM (NYISO), PJM (Neptune), PJM (Linden) & PJM (Hudson)
            # No. 2 - US-PJM->TBD1
                # PJM (DOM - CPL Retail Energy East) to be determined, but a sub-state geometry
            # No. 3 - US-PJM->US_TBD2
                #PJM (AEP - Duke Energy) to be determined but likely a parser for NC & SC
            # No. 4 - US-PJM->US_TBD3
                # PJM (AEP - CPL Retail Energy West) to be determined, but a sub-state geometry
            # No. 5 - US-PJM->US-TVA
                # PJM (AEP - TVA), will be parser for US_TVA (Tennessee)
            # No. 6 - US-PJM->TBD4
                # PJM (EKPC - Louisville G&E Co), Kentucky geometry to be split in half --> Louisville is at the split
            # No. 7 - US-PJM->TBD5
                # PJM (DEOK - Ohio Valley Electric Co), to be determined, but likely a sub-state geometry
            # No. 8 - US-PJM->US-MISO
                # PJM (MISO) - MISO will be multiple state geometry with parser
            # No. 9 - US-NY->US-MISO
                # "Lake Erie Loop" is balancing for NYISO<-->IESO & MISO<-->IESO.
                # + MW for MISO (Michigan) equal - MW for NYISO,
                # passes thru PJM only, no balancing in PJM needed